### ISSUE TITLE
Panic while rsync tries to access snapshots inside OpenVZ container

### ISSUE THREAD

**Issue by lazy404** [zfsonlinux/zfs#3887]
zfs 0.6.5.2

usermode helper probably doesn't know the correct absolute location of the .zfs dir and mount fails,

[  237.148072] Kernel panic - not syncing: avl_find() succeeded inside avl_add()
[  237.155258] Pid: 6310, comm: rsync veid: 202 Tainted: P           ---------------    2.6.32-openvz-042stab111.11-amd64 #3
[  237.166261] Call Trace:
[  237.168764]  [<ffffffff8152d031>] ? panic+0xa7/0x167
[  237.173773]  [<ffffffffa0273780>] ? avl_remove+0x0/0x240 [zavl]
[  237.179772]  [<ffffffffa035d0aa>] ? zfsctl_snapshot_add+0x3a/0x50 [zfs]
[  237.186454]  [<ffffffffa035e75b>] ? zfsctl_snapshot_mount+0x3cb/0x460 [zfs]
[  237.193483]  [<ffffffffa0350f04>] ? zap_get_leaf_byblk+0xe4/0x2c0 [zfs]
[  237.200166]  [<ffffffffa038f8c0>] ? zpl_snapdir_automount+0x10/0x30 [zfs]
[  237.207001]  [<ffffffff811bdc9e>] ? follow_managed+0x17e/0x330
[  237.212878]  [<ffffffff811bdf01>] ? do_lookup+0xb1/0x2b0
[  237.218233]  [<ffffffff81536bfd>] ? mutex_lock+0x1d/0x50
[  237.223588]  [<ffffffff811cb409>] ? dput_nocache+0xc9/0x260
[  237.229203]  [<ffffffff811c0637>] ? __link_path_walk+0x4a7/0x880
[  237.235252]  [<ffffffff811c0a65>] ? path_walk+0x55/0xb0
[  237.240521]  [<ffffffff811c0c1b>] ? filename_lookup+0x6b/0xc0
[  237.246310]  [<ffffffff811b0950>] ? get_empty_filp+0x90/0x1b0
[  237.252099]  [<ffffffff811c20d7>] ? do_filp_open+0xf7/0xce0
[  237.257715]  [<ffffffff810582c3>] ? __wake_up+0x53/0x70
[  237.262987]  [<ffffffff812a3dd7>] ? __strncpy_from_user+0x27/0x60
[  237.269183]  [<ffffffff812a3e41>] ? strncpy_from_user+0x31/0x40
[  237.275204]  [<ffffffff811c1657>] ? getname_flags.part.43+0xc7/0x1d0
[  237.281654]  [<ffffffff811d0905>] ? alloc_fd+0x55/0x140
[  237.286978]  [<ffffffff811ad025>] ? do_sys_open+0x65/0x130
[  237.292559]  [<ffffffff811ad110>] ? sys_open+0x20/0x30
[  237.297829]  [<ffffffff8100b1a2>] ? system_call_fastpath+0x16/0x1b

1444039264   mount; name=vz/backup/private/202@auto-daily-2015-09-07-2300 path=/.zfs/snapshot/auto-daily-2015-09-07-2300
1444039266   txg=134082 quiesce_txg=134083 sync_txg=134069
1444039266   quiesce done, handing off txg 134082
1444039266   txg=134082 quiesce_txg=134083 sync_txg=134069
1444039266   waiting; tx_synced=134082 waiting=134069 dp=ffff88106dc42800

interesting thing is that the cd seems to be successfull

root@lamp:/# cd /.zfs/snapshot/auto-daily-2015-09-07-2300/
root@lamp:/.zfs/snapshot/auto-daily-2015-09-07-2300# ls
aquota.group  aquota.user  bin  boot  dev  etc  home  lib  lib64  lost+found  media  mnt  opt  proc  root  run  sbin  selinux  srv  sys  tmp  usr  var

trying to access ather snapshots fails with a panic as above

strace rsync -aP /.zfs /test
open(".zfs/snapshot", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 3
getdents(3, /* 14 entries */, 32768)    = 624
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
mmap(NULL, 266240, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7fc2705c2000
lstat(".zfs/snapshot/auto-weekly-2015-09-13-2330", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-daily-2015-09-13-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-weekly-2015-09-06-2330", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-daily-2015-09-12-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-daily-2015-09-16-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-daily-2015-09-08-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-daily-2015-09-10-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-daily-2015-09-14-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-daily-2015-09-09-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-daily-2015-09-11-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
lstat(".zfs/snapshot/auto-daily-2015-09-15-2300", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
getdents(3, /* 0 entries */, 32768)     = 0
close(3)                                = 0
open(".zfs/snapshot/auto-daily-2015-09-07-2300", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 3
getdents(3, /* 28 entries */, 32768)    = 776
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/.zfs", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 100rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/.zfs") failed: Object is remote (66)) = 100
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/.cpt_hardlink_dir_a920e4ddc233afddc9fb53d26c392319", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 146rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/.cpt_hardlink_dir_a920e4ddc233afddc9fb53d26c392319") failed: Object is remote (66)) = 146
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/srv", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/srv") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/opt", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/opt") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/proc", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 100rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/proc") failed: Object is remote (66)) = 100
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/root", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 100rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/root") failed: Object is remote (66)) = 100
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/media", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 101rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/media") failed: Object is remote (66)) = 101
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/sys", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/sys") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/selinux", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 103rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/selinux") failed: Object is remote (66)) = 103
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/tmp", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/tmp") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/mnt", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/mnt") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/lib64", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 101rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/lib64") failed: Object is remote (66)) = 101
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/test_zfs", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 104rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/test_zfs") failed: Object is remote (66)) = 104
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/boot", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 100rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/boot") failed: Object is remote (66)) = 100
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/lost+found", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 106rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/lost+found") failed: Object is remote (66)) = 106
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/var", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/var") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/dev", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/dev") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/aquota.user", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 107rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/aquota.user") failed: Object is remote (66)) = 107
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/etc", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/etc") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/run", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/run") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/bin", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/bin") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/usr", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/usr") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/sbin", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 100rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/sbin") failed: Object is remote (66)) = 100
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/lib", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 99rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/lib") failed: Object is remote (66)) = 99
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/aquota.group", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 108rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/aquota.group") failed: Object is remote (66)) = 108
write(2, "\n", 1
)                       = 1
lstat(".zfs/snapshot/auto-daily-2015-09-07-2300/home", 0x7ffc70427220) = -1 EREMOTE (Object is remote)
write(2, "rsync: readlink_stat(\"/.zfs/snap"..., 100rsync: readlink_stat("/.zfs/snapshot/auto-daily-2015-09-07-2300/home") failed: Object is remote (66)) = 100
write(2, "\n", 1
)                       = 1
getdents(3, /* 0 entries */, 32768)     = 0
close(3)                                = 0
open(".zfs/snapshot/auto-daily-2015-09-08-2300", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC

**Comment by behlendorf** [comment]
@lazy404 could you verify the patch in #3890 prevents the panic.  It's not a proper long term fix for handling auto-mounts in multiple namespaces but how to properly handle that still needs to be thought through carefully.

**Comment by lazy404** [comment]
@behlendorf still panics, mayby it has to do with mount being unsuccessfull ?

[  240.328670] Kernel panic - not syncing: avl_find() succeeded inside avl_add()
[  240.335857] Pid: 5410, comm: mc veid: 202 Tainted: P           ---------------    2.6.32-openvz-042stab111.12-amd64 #1
[  240.346598] Call Trace:
[  240.349101]  [<ffffffff81538f20>] ? panic+0xa7/0x16f
[  240.354110]  [<ffffffffa029a19d>] ? avl_find+0x5d/0xb0 [zavl]
[  240.359899]  [<ffffffffa029a19d>] ? avl_find+0x5d/0xb0 [zavl]
[  240.365689]  [<ffffffffa029a780>] ? avl_remove+0x0/0x240 [zavl]
[  240.371689]  [<ffffffffa03b7d79>] ? zfsctl_snapshot_add+0x37/0x3b [zfs]
[  240.378377]  [<ffffffffa037a76a>] ? zfsctl_snapshot_mount+0x42a/0x470 [zfs]
[  240.385412]  [<ffffffffa03ab880>] ? zpl_snapdir_automount+0x10/0x30 [zfs]
[  240.392245]  [<ffffffff811c6082>] ? follow_managed+0x222/0x2e0
[  240.398119]  [<ffffffff811c61f5>] ? do_lookup+0xb5/0x270
[  240.403474]  [<ffffffff811c71aa>] ? __link_path_walk+0x8da/0x10b0
[  240.409610]  [<ffffffff811c7bea>] ? path_walk+0x6a/0xe0
[  240.414879]  [<ffffffff811c7dfb>] ? filename_lookup+0x6b/0xc0
[  240.420668]  [<ffffffff811c8f49>] ? user_path_at+0x59/0xa0
[  240.426198]  [<ffffffff811b62bf>] ? vfs_read+0x12f/0x1a0
[  240.431554]  [<ffffffff811b35b9>] ? sys_chdir+0x39/0xf0
[  240.436822]  [<ffffffff8100b1a2>] ? system_call_fastpath+0x16/0x1b

**Comment by ryao** [comment]
@lazy404 Is it possible that you loaded the unpatched module?

**Comment by lazy404** [comment]
@ryao unlikelly, I have recompiled it 2 times and rechecked

**Comment by behlendorf** [comment]
@lazy404 thanks for testing and re-verifying.  I had another thought about this, does you system by chance contain multiple pools?  If so that could explain things.  While all the names are unique (due to the pool name being includes) the objset ids are not and that could lead to this.

**Comment by lazy404** [comment]
@behlendorf single pool, it panics at second mount attempt on when I try to enter a difrent snapshot

I will try to add some debug messages and check if hardcoding correct mount patch will help

**Comment by behlendorf** [comment]
Thanks getting a clear reproducer or any additional debugging would be helpful.  It would be straight forward to catch the error and handle it but I'd really like to understand exactly under what circumstances this can happen.

**Comment by christoph-h** [comment]
I guess I also ran into the same issue. I don't have a lot of information since it happened on a production machine with quite some data where did not want to do a lot of experiments. Of course I'm open to suggestions.

I'm running two Proxmox VE 3.4 servers, each of which has 5 OpenVZ containers running (Proxmox 3.4 is based on Wheezy 7.8, but they ship with their own kernel). Both servers have a single root pool setup, that was setup during the proxmox installation. I added the SSDs for cache and log:
```
  pool: rpool
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
	still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
	the pool may no longer be accessible by software that does not support
	the features. See zpool-features(5) for details.
  scan: scrub repaired 0 in 5h57m with 0 errors on Fri Oct  9 08:16:19 2015
config:

	NAME                                                  STATE     READ WRITE CKSUM
	rpool                                                 ONLINE       0     0     0
	  raidz2-0                                            ONLINE       0     0     0
	    ata-WDC_WD20EURX-63T0FY0_WD-WCC4M0739211-part2    ONLINE       0     0     0
	    ata-WDC_WD20EURX-63T0FY0_WD-WCC4M0749067-part2    ONLINE       0     0     0
	    ata-WDC_WD20EURX-64HYZY0_WD-WMC4M1750616-part2    ONLINE       0     0     0
	    ata-WDC_WD20EURX-63T0FY0_WD-WCC4M0713700-part2    ONLINE       0     0     0
	    ata-WDC_WD20EURX-64HYZY0_WD-WMC4M1655808-part2    ONLINE       0     0     0
	    ata-WDC_WD20EURX-63T0FY0_WD-WCC4M0713713-part2    ONLINE       0     0     0
	logs
	  mirror-1                                            ONLINE       0     0     0
	    ata-KINGSTON_SV300S37A60G_50026B773B048C59-part2  ONLINE       0     0     0
	    ata-KINGSTON_SV300S37A60G_50026B773B048C5A-part2  ONLINE       0     0     0
	cache
	  ata-KINGSTON_SV300S37A60G_50026B773B048C59-part3    ONLINE       0     0     0
	  ata-KINGSTON_SV300S37A60G_50026B773B048C5A-part3    ONLINE       0     0     0

errors: No known data errors
```

Regarding this bug, both servers always behaved the same, so they both gave me a kernel panic (Sorry, I only got IPMI screenshots):
![zfs_bug__dc_reporduced](https://cloud.githubusercontent.com/assets/6417771/10540164/cf36239e-7405-11e5-9fbe-8557e47cd88d.png)



Proxmox 3.4 has its own ZFS packages included in their repos ( http://download.proxmox.com/debian/dists/wheezy/pve-no-subscription/binary-amd64/ ). I guess those are useful versions:
```
pve-kernel-2.6.32-42-pve:amd64
libnvpair1:amd64 0.6.5-1~wheezy
libuutil1:amd64 0.6.5-1~wheezy
libzfs2:amd64 0.6.5-1~wheezy
libzpool2:amd64 0.6.5-1~wheezy
spl:amd64 0.6.5-1~wheezy
zfs-initramfs:amd64 0.6.5-1~wheezy
zfsutils:amd64 0.6.5-1~wheezy
```
(I haven't checked which 6.5.x ZFS version this correspsonds to, and since I downgraded to 6.4 I cant check now)

The bug always occured around 5-6am, so it seems to be some system's cron job thing. Right before the kernel panic, there is an OpenVZ container 116 mentioned. I guess it is linked to this cronjob ( /etc/cron.daily/mlocate ) 
```
#! /bin/bash

set -e

[ -x /usr/bin/updatedb.mlocate ] || exit 0

if which on_ac_power >/dev/null 2>&1; then
    ON_BATTERY=0
    on_ac_power >/dev/null 2>&1 || ON_BATTERY=$?
    if [ "$ON_BATTERY" -eq 1 ]; then
        exit 0
    fi
fi

##

LOCKFILE="/var/lib/mlocate/daily.lock"

trap "rm -f $LOCKFILE" EXIT

if [ -e "$LOCKFILE" ]; then
    echo >&2 "Warning: $LOCKFILE present, not running updatedb."
    exit 1
else
    touch "$LOCKFILE"
fi

##

# See ionice(1)
if [ -x /usr/bin/ionice ] &&
    /usr/bin/ionice -c3 true 2>/dev/null; then
    IONICE="/usr/bin/ionice -c3"
fi

$IONICE /usr/bin/updatedb.mlocate
```
Going through the script, I executed
```
/usr/bin/ionice -c3 /usr/bin/updatedb.mlocate
```
and I could reproduce the panic, as shown above.
However, neither of the following commands caused a panic.
```
# no panic here
/usr/bin/ionice -c3 /bin/sleep 3

# this command does not cause a panic either, but it also doesn't return the shell prompt (don't know if that is normal)
/usr/bin/updatedb.mlocate
```

Here are some versions:
```
/usr/bin/updatedb.mlocate --version
updatedb (mlocate) 0.23.1
Copyright (C) 2007 Red Hat, Inc. All rights reserved.
This software is distributed under the GPL v.2.

This program is provided with NO WARRANTY, to the extent permitted by law.
```
```
/usr/bin/ionice --version
ionice from util-linux 2.20.1
```









I'm pretty sure this bug was introduced between ZFS 6.4 and 6.5. There is also a bit more of a background: Before seeing this issue with Proxmox 3.4, I used to get this issue when using Proxmox 3.3 which does not have its own ZFS packages (so I used the ZFSonLinux repo). There was only one data ZFS pool, so the proxmox system was not install within the ZFS (however, the filesystem root of the containers were in the ZFS pool). I started getting the kernel panics after the update to ZFS 6.5. Again, it happened always around 5-6am, so I guess it is the same things going on. Then, I completely wiped the machines and installed proxmox 3.4 since it has its own ZFS repo and I also switched to a root ZFS pool (that's what I desribed above).


After seeing the kernel panic on my rather freshly installed Proxmox 3.4 (by coincidence, the Proxmox repos just updated to 6.5 a few days after I installed proxmox 3.4), I downgraded to ZFS 6.4 (still in the proxmox repo), but I am still getting these kernel panics.


So in summary, that's what I can say:
 - With ZFS 6.4 I never had this issue, so it was introduced in ZFS 6.5. However, once ZFS 6.5 was installed, a downgrade to 6.4 does not cure the problem.
 - I also have OpenVZ containers running.
 - It can be reproduced by runnning ```/usr/bin/ionice -c3 /usr/bin/updatedb.mlocate```

I hope this is of any help! Please let me know if you need more info.

**Comment by behlendorf** [comment]
This looks like a duplicate of #3786 which is believed to be understood.  We'll see if we can't put a patch together for the next point release.  Thanks for the detailed report.

**Comment by behlendorf** [comment]
This should now be addressed in master and these fixes will appear in the next point release.  @tuxoko provided patches for both known ways this could occur.  If people impacted by this issue have time to independently verify the fix which was applied to master that would be appreciated.

24ef51f Use spa as key besides objsetid for snapentry
d287880 Fix snapshot automount behavior when concurrent or fail

**Comment by christoph-h** [comment]
I just got the latest ZFS via the proxmox repos. And although this issue is already closed, I still wanted to confirm that this bug is now fixed. I can run /usr/bin/updatedb.mlocate inside the OpenVZ container without any issues.
Thanks a lot for your great work!
